src/functions.c

```diff
@@ -0,0 +1,447 @@
+        dictSdsDup,        /* key dup */
+        NULL,              /* val dup */
+        dictSdsKeyCompare, /* key compare */
+        dictSdsDestructor, /* key destructor */
+        NULL,              /* val destructor */
+        NULL               /* allow to expand */
+};
+
+/* Dictionary of engines */
+static dict *engines = NULL;
+
+/* Functions Ctx.
+ * Contains the dictionary that map a function name to
+ * function object and the cache memory used by all the functions */
+static functionsCtx *functions_ctx = NULL;
+
+static size_t functionMallocSize(functionInfo *fi) {
+    return zmalloc_size(fi) + sdsZmallocSize(fi->name)
+            + (fi->desc ? sdsZmallocSize(fi->desc) : 0)
+            + sdsZmallocSize(fi->code)
+            + fi->ei->engine->get_function_memory_overhead(fi->function);
+}
+
+/* Dispose function memory */
+static void engineFunctionDispose(functionInfo *fi) {
+    sdsfree(fi->code);
+    sdsfree(fi->name);
+    if (fi->desc) {
+        sdsfree(fi->desc);
+    }
+    engine *engine = fi->ei->engine;
+    engine->free_function(engine->engine_ctx, fi->function);
+    zfree(fi);
+}
+
+/* Free function memory and detele it from the functions dictionary */
+static void engineFunctionFree(functionInfo *fi, functionsCtx *functions) {
+    functions->cache_memory -= functionMallocSize(fi);
+
+    dictDelete(functions->functions, fi->name);
+    engineFunctionDispose(fi);
+}
+
+/* Clear all the functions from the given functions ctx */
+void functionsCtxClear(functionsCtx *functions_ctx) {
+    dictIterator *iter = dictGetIterator(functions_ctx->functions);
+    dictEntry *entry = NULL;
+    while ((entry = dictNext(iter))) {
+        functionInfo *fi = dictGetVal(entry);
+        engineFunctionDispose(fi);
+    }
+    dictReleaseIterator(iter);
+    dictEmpty(functions_ctx->functions, NULL);
+    functions_ctx->cache_memory = 0;
+}
+
+/* Free the given functions ctx */
+void functionsCtxFree(functionsCtx *functions_ctx) {
+    functionsCtxClear(functions_ctx);
+    dictRelease(functions_ctx->functions);
+    zfree(functions_ctx);
+}
+
+/* Swap the current functions ctx with the given one.
+ * Free the old functions ctx. */
+void functionsCtxSwapWithCurrent(functionsCtx *new_functions_ctx) {
+    functionsCtxFree(functions_ctx);
+    functions_ctx = new_functions_ctx;
+}
+
+/* return the current functions ctx */
+functionsCtx* functionsCtxGetCurrent() {
+    return functions_ctx;
+}
+
+/* Create a new functions ctx */
+functionsCtx* functionsCtxCreate() {
+    functionsCtx* ret = zmalloc(sizeof(functionsCtx));
+    ret->functions = dictCreate(&functionDictType);
+    ret->cache_memory = 0;
+    return ret;
+}
+
+/*
+ * Register a function info to functions dictionary
+ * 1. Set the function client
+ * 2. Add function to functions dictionary
+ * 3. update cache memory
+ */
+static void engineFunctionRegister(functionInfo *fi, functionsCtx *functions) {
+    dictAdd(functions->functions, fi->name, fi);
```

Comment:
Although `functionsCreateWithFunctionCtx` has already deleted the key, I think it's safer to add an assertion.
```suggestion
    assert(dictAdd(functions->functions, fi->name, fi) == DICT_OK);
```

Reply:
I do not like putting code that should run inside an assert (in case we want to run without asserts someday), will save the return value and assert it.

Reply:
Yes, it saves a line of code when writing in assert if we don't need the return value.

Reply:
I mean that I want to save the return value in a variable and then assert, this way if someone compiles without asserts it will still work:
```
int res = dictAdd(functions->functions, fi->name, fi);
assert(res == DICT_OK);
```